From 68d275c7558c9a315829e7119629c111712fe7d4 Mon Sep 17 00:00:00 2001
From: Min He <min.he@intel.com>
Date: Wed, 25 Apr 2018 20:20:33 +0800
Subject: [PATCH 1040/1240] drm/i915/gvt: set digital bit in the virtual EDID

Sometimes physical monitor will not report the digital bit in its EDID,
which will cause GVT-g guest fail to enumerate the virtual HDMI monitor.
To avoid such issue, we will set the digital bit in virtual EDID we
present to guest.

This patch fixed the display issue of GVT-g guest on the APL NUC.

V2: Fixed the calculation of the sum.

Signed-off-by: Min He <min.he@intel.com>
Acked-by: Zhao Yakui <yakui.zhao@intel.com>
---
 drivers/gpu/drm/i915/gvt/display.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/gpu/drm/i915/gvt/display.c b/drivers/gpu/drm/i915/gvt/display.c
index 7ab1587..a78266d 100644
--- a/drivers/gpu/drm/i915/gvt/display.c
+++ b/drivers/gpu/drm/i915/gvt/display.c
@@ -301,6 +301,7 @@ static int setup_virtual_monitor(struct intel_vgpu *vgpu, int port_num,
 {
 	struct intel_vgpu_port *port = intel_vgpu_port(vgpu, port_num);
 	int valid_extensions = 1;
+	struct edid *tmp_edid = NULL;
 
 	if (WARN_ON(resolution >= GVT_EDID_NUM))
 		return -EINVAL;
@@ -326,6 +327,17 @@ static int setup_virtual_monitor(struct intel_vgpu *vgpu, int port_num,
 				virtual_dp_monitor_edid[resolution],
 				EDID_SIZE);
 
+	/* Sometimes the physical display will report the EDID with no
+	 * digital bit set, which will cause the guest fail to enumerate
+	 * the virtual HDMI monitor. So here we will set the digital
+	 * bit and re-calculate the checksum.
+	 */
+	tmp_edid = ((struct edid *)port->edid->edid_block);
+	if (!(tmp_edid->input & DRM_EDID_INPUT_DIGITAL)) {
+		tmp_edid->input += DRM_EDID_INPUT_DIGITAL;
+		tmp_edid->checksum -= DRM_EDID_INPUT_DIGITAL;
+	}
+
 	port->edid->data_valid = true;
 
 	if (is_dp) {
-- 
2.7.4

